File: implementations/jsu-py/jsonschema_benchmark.py
#! /bin/env python

import argparse
import json
import logging
import sys
import time

import schema

def jsonschema_benchmark():
    """Run JSON Schema Benchmark on a an instances files"""

    logging.basicConfig()
    log = logging.getLogger("jsb")

    ap = argparse.ArgumentParser()
    ap.add_argument("--debug", action="store_true", help="set debug")
    ap.add_argument("--clock", "-C", default="r", help="set clock")
    ap.add_argument("values", nargs="*", help="JSONL files")
    args = ap.parse_args()

    log.setLevel(logging.DEBUG if args.debug else logging.INFO)

    clock: int
    if args.clock in ("r", "realtime"):
        clock = time.CLOCK_REALTIME
    elif args.clock in ("m", "monotonic"):
        clock = time.CLOCK_MONOTONIC
    elif args.clock in ("p", "process"):
        clock = time.CLOCK_PROCESS_CPUTIME_ID
    elif args.clock in ("t", "thread"):
        clock = time.CLOCK_THREAD_CPUTIME_ID
    else:
        raise Exception(f"unexpected clock spec: {args.clock}")

    errors = 0
    schema.check_model_init()
    # validator for root model
    checker = schema.check_model_fun("")

    # load all files into values
    values = []
    for fn in args.values:
        log.debug(f"considering file {fn}")

        # load jsonl data
        with open(fn) as f:
            values += [json.loads(r) for r in f]

    # overhead estimation in µs
    count = 0
    overhead_start = time.clock_gettime(clock)
    for v in values:
        count += 1
    overhead_delay = 1_000_000.0 * (time.clock_gettime(clock) - overhead_start)

    # cold run in µs
    cold_start = time.clock_gettime(clock)
    for v in values:
        if not checker(v, None, None):
            errors += 1
    cold_delay = 1_000_000.0 * (time.clock_gettime(clock) - cold_start)
    log.debug(f"cold delay: {cold_delay:.03f} µs")

    # warmup runs to trigger a potential JIT?
    warmup = min(1000, 1 + int(10_000_000.0 / cold_delay))
    log.debug(f"warmup loop: {warmup}")
    while warmup > 0:
        for v in values:
            checker(v, None, None)
        warmup -= 1

    # warm run in µs
    start = time.clock_gettime(clock)
    for v in values:
        checker(v, None, None)
    delay = 1_000_000.0 * (time.clock_gettime(clock) - start)

    # result for humans
    print(f"py validation: pass={len(values) - errors} fail={errors} "
          f"{delay:.03f} µs [{overhead_delay:.03f} µs]", file=sys.stderr)

    # cold-run-ns,warm-run-ns
    print(f"{int(1000 * cold_delay + 0.5)},{int(1000 * delay + 0.5)}")

    schema.check_model_free()
    sys.exit(1 if errors else 0)

if __name__ == "__main__":
    jsonschema_benchmark()

File: implementations/jsu-js/jsonschema_benchmark.js
import { parseArgs } from 'node:util'
import fs from 'node:fs/promises'

import { jm_set_rx } from 'json_model_runtime'
import { createRequire } from 'node:module'
const require = createRequire(import.meta.url)

import { check_model_init, check_model_map, check_model_free } from './schema.js'

export default async function main()
{
    const options = {
      // TODO --help --list
      'debug': { type: 'boolean', short: 'D' },
      're2': { type: 'boolean' },
      'regexp': { type: 'boolean' },
    }

    const args = parseArgs({options, allowPositionals: true})

    // reset regular expression engine
    if (args.values.regexp)
        jm_set_rx(RegExp)
    if (args.values.re2)
        jm_set_rx(require('re2'))

    let debug = args.values.debug
    let errors = 0
    check_model_init()
    const checker = check_model_map.get("")

    // load files contents
    const values = []
    for (const fname of args.positionals)
    {
        const data = await fs.readFile(fname, {encoding: 'UTF-8'})
        values.push(...data.split("\n").slice(0, -1).map(s => JSON.parse(s)))
    }

    // overhead estimation
    let count = 0
    const overhead_start = performance.now()
    for (const v of values)
        if (v !== null)
            count++
    const overhead_delay = performance.now() - overhead_start  // ms

    // cold run
    if (debug)
        console.error("cold run")
    const cold_start = performance.now()
    for (const v of values)
        if (!checker(v, "", null))
            errors++
    const cold_delay = performance.now() - cold_start  // ms

    // warm-up so as to trigger JIT
    const WARMUP_MAX_TIME = 10.0  // seconds
    const WARMUP_ITERATIONS = 1000  // unless too long
    const max_iterations = Math.ceil(WARMUP_MAX_TIME * 1000.0 / cold_delay)
    const warmup_iterations = Math.min(WARMUP_ITERATIONS, max_iterations)

    if (debug)
        console.error("warmup loop ${warmup_iterations}")

    for (let i = 0; i < warmup_iterations; i++)
        for (const v of values)
            checker(v, '', null)

    // warm run
    if (debug)
        console.error("hot run")
    const start = performance.now()
    for (const v of values)
        checker(v, "", null)
    const delay = performance.now() - start  // ms

    console.error(`js validation: pass=${values.length - errors} fail=${errors}`,
                  `${(1000.0 * delay).toFixed(3)} µs [${(1000.0 * overhead_delay).toFixed(3)} µs]`)

    console.log((1000000.0 * cold_delay).toFixed(0) + ',' + (1000000.0 * delay).toFixed(0))

    check_model_free()
    process.exit(errors ? 1 : 0)
}

main()
